discretisedfield/tools/tools.py

```diff
@@ -105,52 +105,59 @@ def topological_charge_density(field, /, method="continuous"):
         msg = f"Cannot compute topological charge density for {field.nvdim=} field."
         raise ValueError(msg)
 
-    if not field.mesh.attributes["isplane"]:
-        msg = (
-            "The field must be sliced before the topological "
-            "charge density can be computed."
+    if field.mesh.region.ndim != 2:
+        raise ValueError(
+            "The topological charge density can only be computed on fields with 2"
+            f" spatial dimensions, not {field.mesh.region.ndim=}."
         )
-        raise ValueError(msg)
 
     if method not in ["continuous", "berg-luescher"]:
         msg = "Method can be either continuous or berg-luescher"
         raise ValueError(msg)
 
-    axis1 = field.mesh.attributes["axis1"]
-    axis2 = field.mesh.attributes["axis2"]
     of = field.orientation  # unit field - orientation field
 
+    # Spatial axes associated with the first and second vector
+    # dimensions.
+    axis1 = field.vdim_mapping[field.vdims[0]]
+    axis2 = field.vdim_mapping[field.vdims[1]]
```

Comment:
We should consider adding some checks here as `vdim_mapping` may be empty or return axes that are not part of `region.dims`.

Reply:
If I move `axis1_idx = field.mesh.region._dim2index(axis1)` directly below then this should fix this as it has inbuilt error checking. Would you be happy with this?

Reply:
Yes, I think that should be sufficient.